### PR TITLE
Use the CertificateAuthority store for finding intermediate certs in macOS

### LINKED ISSUE #48207: certificates added to user CertificateAuthority store are ignored by X509Chain.Build on macOS
This is subset of  #48017 as the underlying issue seems different.
I basically have fragment like 

```c#
    using (X509Store store = new X509Store(StoreName.CertificateAuthority, StoreLocation.CurrentUser))
    {
        // add chain certificate so we can construct chain since there is no way how to pass intermediates directly.
        store.Open(OpenFlags.ReadWrite);
        store.AddRange(clientChain);
        store.Close();
    }
    
    var chain = new X509Chain();
    ....
    bool chainStatus = chain.Build(clientCertificate);
```
It works on Linux & Windows but it fails consistently on macOS (10.15) 
Interestingly, it works if I use `StoreName.My` instead.

It seems like that store does not have standard OS KeyChain mapping and that is reason why the chain does not build correctly. If that is true, it would be nice if we can feed reference to the keychain or it's collection automaticaly so validation and chain building works consistently across platforms. 
In SslStream this primarily impacts cases with client certificates.

### PR DESCRIPTION
**Pull request by vcsjones**
If the `StoreName.CertificateAuthority` store exists, we can use it as a store to find extra intermediate certificates when building an X.509 chain.

Closes #48207

### CODE REVIEW

File: src/libraries/System.Security.Cryptography.X509Certificates/tests/DynamicChainTests.cs [hunk 1]
```diff
@@ -832,6 +833,56 @@ public static void PolicyConstraints_Mapped()
             }
         }
 
+        [Fact]
+        [OuterLoop("Modifies a shared trust store.")]
```
Comment (vcsjones) [inline src/libraries/System.Security.Cryptography.X509Certificates/tests/DynamicChainTests.cs:837]: I think it makes sense for this test to be outer loop, but maybe I am being too cautious.

The test attempts to be clean up before and after itself. The intermediate certificates get a marker extension for the test, and cleaned up before and after test execution.

File: src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/ChainPal.Apple.cs [hunk 2]
```diff
@@ -38,9 +39,16 @@ internal SecTrustChainPal()
             X509Certificate2Collection? customTrustStore,
             X509ChainTrustMode trustMode)
         {
+            _certificateAuthorityStore?.Dispose();
+            _certificateAuthorityStore = OpenCertificateAuthorityStore();
```
Comment (vcsjones) [inline src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/ChainPal.Apple.cs:43]: My biggest question here is the lifetime of the X509Store and its certificates are managed correctly here. I think this is correct, but, could use some input here. My concerns largely are making sure the cert handles are cleaned up correctly, and likely not freed before we are done with them.

And of course ideas for tests that prove that out are certainly welcome.

**Comment by vcsjones** [comment]
It looks like the failing test is legitimate. It's failing because previously, the test was getting skipped. Now that X509Chain is able to use `CertificateAuthority`, the test is not getting skipped anymore.

https://github.com/dotnet/runtime/blob/04e65d2b871211832f9999256435787c727afd0f/src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamNetworkStreamTest.cs#L879-L881

But, the test is now failing for another reason.

I'll try to fix it, but this is a little out of my expertise.

**Comment by wfurt** [comment]
I think there is some race condition in processing the certificates @vcsjones. The tests failing in CI were passing fine on my local macOS.

**Comment by vcsjones** [comment]
@wfurt hm, okay. I will revive this and see if I can figure out what's happening. Thanks for looking!

**Pull request closed without merging**